libslide/src/math/gcd_poly_zz.rs

```diff
@@ -0,0 +1,430 @@
+            heu = gcd(f_xi.abs() as usize, g_xi.abs() as usize) as isize;
+            cff = f_xi / heu;
+            cfg = g_xi / heu;
+
+            heu_poly = gcd_interpolate(heu, xi);
+            heu_poly = prim_poly(&heu_poly);
+
+            let (_cff2, r) = poly_div(&f, &heu_poly);
+            if r.is_empty() {
+                let (_cfg2, r) = poly_div(&g, &heu_poly);
+                if r.is_empty() {
+                    heu_poly = poly_mul_scalar(&heu_poly, gcd_val);
+                    return Some((heu_poly, _cff2, _cfg2));
+                }
+            }
+            cff_poly = gcd_interpolate(cff, xi);
+            let (mut heu_poly, r) = poly_div(&f, &cff_poly);
+            if r.is_empty() {
+                let (_cfg2, r) = poly_div(&g, &heu_poly);
+                if r.is_empty() {
+                    heu_poly = poly_mul_scalar(&heu_poly, gcd_val);
+                    return Some((heu_poly, cff_poly, _cfg2));
+                }
+            }
+            cfg_poly = gcd_interpolate(cfg, xi);
+            let (mut heu_poly, r) = poly_div(&g, &cfg_poly);
+
+            if r.is_empty() {
+                let (_cff2, r) = poly_div(&f, &heu_poly);
+                if r.is_empty() {
+                    heu_poly = poly_mul_scalar(&heu_poly, gcd_val);
+                    return Some((heu_poly, _cff2, cfg_poly));
+                }
+            }
+        }
+        xi = 73794 * xi * ((xi as f64).sqrt().sqrt() as isize)
+    }
+    None
+}
+
+#[cfg(feature = "benchmark-internals")]
+pub fn _gcd_poly_zz_heu(f: &PolyPtrArg, g: &PolyPtrArg) {
+    gcd_poly_zz_heu(f.to_vec(), g.to_vec());
+}
+
+#[allow(unused)]
+fn max_norm(p: &PolyPtrArg) -> usize {
+    p.iter()
+        .map(|&n| n.abs())
+        .fold(0, |cur_max, n| max(cur_max, n as usize))
+}
+
+#[allow(unused)]
+fn lc(p: &PolyPtrArg) -> isize {
+    *p.get(0).unwrap_or(&0)
+}
+
+#[allow(unused)]
+fn eval(p: &PolyPtrArg, x: isize) -> isize {
+    p.iter().fold(0, |mut res, &n| {
+        res *= x;
+        res + n
+    })
+}
+
+#[allow(unused)]
+fn poly_coeffs_gcd(p: &PolyPtrArg) -> usize {
+    // finds the gcd of the polynomials coeffs
+    if p.is_empty() {
+        return 0;
+    }
+    let mut cont: usize = p[0] as usize;
+    for i in p {
+        cont = gcd(cont, *i as usize);
+        if cont == 1 {
+            break;
+        }
+    }
+    cont as usize
+}
+
+#[allow(unused)]
+fn poly_extract_common(f: &PolyPtrArg, g: &PolyPtrArg) -> (isize, Poly, Poly) {
+    // extracts 'Common gcd content' from polys
+    let mut res_f = f.to_vec();
+    let mut res_g = g.to_vec();
+    let fc = poly_coeffs_gcd(&f);
+    let gc = poly_coeffs_gcd(&g);
+    let gcd = gcd(fc, gc) as isize;
+    if gcd != 1 {
+        res_f = poly_div_scalar(&f, gcd);
+        res_g = poly_div_scalar(&g, gcd);
+    }
+    (gcd, res_f, res_g)
+}
+
+#[allow(unused)]
+// this assumes [] is 0
+fn trivial_gcd(p: &PolyPtrArg, q: &PolyPtrArg) -> Option<(Poly, Poly, Poly)> {
```

Comment:
please add a doc comment for what it means to have a trivial poly GCD and what the return type is